test/PushPullTests.cpp

```diff
@@ -0,0 +1,460 @@
+      ipcLocation.append(".ipc");
+      return ipcLocation;
+   }
+
+   std::vector<std::shared_ptr<Push>> CreatePushs(int howManyPushs,
+                                                  int howManyPulls,
+                                                  const std::string& location) {
+      std::vector<std::shared_ptr<Push>> toReturn;
+      for (int i = 0; i < howManyPushs ; i ++) {
+         if (howManyPushs == 1 && howManyPulls == 1) {
+            // one to one, use default behavior (connect)
+            LOG(DEBUG) << "one to one push setup";
+            toReturn.push_back(std::make_shared<Push>(location));
+         } else if (howManyPushs > 1) {
+            // many pushes to one pull, use default behavior (connect)
+            LOG(DEBUG) << "many to one push setup";
+            toReturn.push_back(std::make_shared<Push>(location));
+         } else if (howManyPulls > 1) {
+            // many pulls to one push, bind instead of connect
+            LOG(DEBUG) << "one to many push setup";
+            toReturn.push_back(std::make_shared<Push>(location, false));
+         } else if (howManyPushs == 1 && howManyPulls == 0) {
+            LOG(DEBUG) << "one push setup";
+            // one push, use default behavior (connect)
+            toReturn.push_back(std::make_shared<Push>(location));
+         }
+      }
+      return toReturn;
+   }
+   std::vector<std::shared_ptr<Pull>> CreatePulls(int howManyPushs,
+                                                  int howManyPulls,
+                                                  const std::string& location) {
+      std::vector<std::shared_ptr<Pull>> toReturn;
+      for (int i = 0; i < howManyPulls ; i ++) {
+         if (howManyPushs == 1 && howManyPulls == 1) {
+            // one to one, use default behavior (bind)
+            LOG(DEBUG) << "one to one pull setup";
+            toReturn.push_back(std::make_shared<Pull>(location));
+         } else if (howManyPushs > 1) {
+            LOG(DEBUG) << "many to one pull setup";
+            // many pushes to one pull, use default behavior (bind)
+            toReturn.push_back(std::make_shared<Pull>(location));
+         } else if (howManyPulls > 1) {
+            LOG(DEBUG) << "one to many pull setup";
+            // many pulls to one push, connect instead of bind
+            toReturn.push_back(std::make_shared<Pull>(location, true));
+         } else if (howManyPushs == 0 && howManyPulls == 1) {
+            LOG(DEBUG) << "one pull setup";
+            // one pull, use default behavior (bind)
+            toReturn.push_back(std::make_shared<Pull>(location));
+         }
+      }
+      return toReturn;
+   }
+   std::vector<std::thread> ListenWithPulls(std::vector<std::shared_ptr<Pull>>& pulls,
+                                            std::function<void(std::shared_ptr<Pull>)> pullLambda) {
+      std::vector<std::thread> threads;
+      for (auto& pull:pulls) {
+         LOG(DEBUG) << "Creating Pull thread";
```

Comment:
It seems many of these log outputs were added during coding and that's good. Not sure if they add anything now and I would suggest cleaning them up (the whole file)